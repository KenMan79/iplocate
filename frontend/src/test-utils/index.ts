export * from "./test-utils";
export * from "./ipdata";
