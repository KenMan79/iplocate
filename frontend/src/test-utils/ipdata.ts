import { IP } from "types";

export const baseIP: IP = {
  continent: {
    code: "EU",
    geonameId: 6255148,
    names: {
      en: "Europe",
    },
  },
  country: {
    geonameId: 3017382,
    isInEuropeanUnion: true,
    isoCode: "FR",
    names: {
      en: "France",
    },
  },
  traits: {
    isAnonymous: false,
    isAnonymousProxy: false,
    isAnonymousVpn: false,
    isHostingProvider: false,
    isLegitimateProxy: false,
    isPublicProxy: false,
    isResidentialProxy: false,
    isSatelliteProvider: false,
    isTorExitNode: false,
    ipAddress: "2.2.2.2",
    network: "2.2.0.0/18",
  },
  city: {},
  location: {
    accuracyRadius: 500,
    latitude: 48.8582,
    longitude: 2.3387,
    timeZone: "Europe/Paris",
  },
};
