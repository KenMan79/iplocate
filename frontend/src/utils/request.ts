export const request = async <T = any>(
  url: string,
  opts: Partial<RequestInit> = {}
): Promise<T> => {
  const resp = await fetch(url, opts);
  if (!resp.ok) {
    throw new Error(resp.statusText);
  }
  const data = await resp.json();
  return data;
};
