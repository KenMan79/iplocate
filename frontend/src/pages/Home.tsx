import React, { useCallback, useState } from "react";
import styled from "styled-components";
import { CollapsableDialog, Map, IPForm, IPMenu, Loading } from "components";
import { IP } from "types";
import { request } from "../utils";
import { usePersistentState } from "hooks";

const endpoint = `${process.env.REACT_APP_API_URL ?? ""}/ip/search`;

export const Home: React.FC = () => {
  const [currentIp, setCurrentIp] = useState<IP>();
  const [allIps, setAllIps] = usePersistentState<IP[]>("allIps", []);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");
  const [menuExpanded, setMenuExpanded] = useState(true);

  const onSearchIP = async (ip: string) => {
    setLoading(true);
    setError("");
    setAllIps((existing) => existing.filter((e) => e.traits.ipAddress !== ip));
    const params = new URLSearchParams({ ip });
    try {
      const ip = await request<IP>(`${endpoint}?${params}`);
      if (!(ip.location.latitude && ip.location.longitude)) {
        throw new Error("Unable to locate IP");
      }
      setCurrentIp(ip);
      setAllIps((allIps) => [ip, ...allIps]);
      setLoading(false);
    } catch (e) {
      setError(e.message);
      setLoading(false);
    }
  };

  const onRemoveIp = useCallback(
    (ip: IP) => {
      setAllIps((existing) =>
        existing.filter(
          (exist) => exist.traits.ipAddress !== ip.traits.ipAddress
        )
      );
    },
    [setAllIps]
  );

  const onToggleIpVisible = useCallback(
    (ip: IP) => {
      setAllIps((existing) =>
        existing.map((e) => ({
          ...e,
          hidden:
            e.traits.ipAddress === ip.traits.ipAddress ? !e.hidden : e.hidden,
        }))
      );
    },
    [setAllIps]
  );

  const onSetCurrentIP = useCallback(
    (ip: IP) => {
      if (window.matchMedia("(max-width: 450px)").matches) {
        setMenuExpanded(false);
      }
      setCurrentIp(ip);
    },
    [setCurrentIp]
  );

  return (
    <Wrapper>
      {loading && <Loading />}
      <CollapsableDialog
        title="IP Locate"
        collapsed={!!currentIp && allIps.length > 0}
        menuExpanded={menuExpanded}
        setMenuExpanded={setMenuExpanded}
      >
        {error && <ErrorMsg>{error}</ErrorMsg>}
        <IPForm onSubmit={(ip) => onSearchIP(ip)} />
        <IPMenu
          ips={allIps}
          onRemoveIp={onRemoveIp}
          onSetCurrentIp={onSetCurrentIP}
          onToggleIpVisible={onToggleIpVisible}
        />
      </CollapsableDialog>
      <Map
        selectedIP={currentIp}
        onSetSelectedIp={onSetCurrentIP}
        allIPs={allIps}
      />
    </Wrapper>
  );
};

const Wrapper = styled.div`
  flex: 1;
  overflow-y: hidden;
  position: relative;
  width: 100%;
  height: 100%;
`;

const ErrorMsg = styled.p`
  margin: 0.5rem 0;
  color: red;
`;
