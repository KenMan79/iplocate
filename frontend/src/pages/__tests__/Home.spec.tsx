import React from "react";
import { render, baseIP, fireEvent, wait } from "test-utils";
import { Home } from "../Home";
import { request } from "utils/request";

const otherIP = {
  ...baseIP,
  traits: {
    ...baseIP.traits,
    ipAddress: "9.9.9.9",
  },
};

jest.mock("components/Map", () => ({
  Map: () => <div />,
}));
jest.mock("utils/request", () => ({
  request: jest.fn(),
}));
const mockRequest = request as jest.Mock;

jest.spyOn(Storage.prototype, "setItem");
const setItemMock = localStorage.setItem as jest.Mock;

jest.spyOn(Storage.prototype, "getItem");
const getItemMock = localStorage.getItem as jest.Mock;

let oldMatchMedia: typeof window.matchMedia;
const mediaQueryMock = jest.fn();

beforeAll(() => {
  oldMatchMedia = window.matchMedia;
  window.matchMedia = mediaQueryMock;
});

beforeEach(() => {
  getItemMock.mockReturnValue(null);
});

afterEach(() => {
  mockRequest.mockReset();
  setItemMock.mockReset();
  getItemMock.mockReset();
  mediaQueryMock.mockReset();
});

afterAll(() => {
  jest.restoreAllMocks();
  window.matchMedia = oldMatchMedia;
});

const renderAndSearch = async (value: string) => {
  const { getByLabelText, ...rest } = render(<Home />);
  const inp = getByLabelText("IP Address input");
  fireEvent.change(inp, {
    target: { value },
  });
  await wait(() => expect(inp).toHaveValue(value));
  fireEvent.click(getByLabelText("Search IP Address"));
  await wait(() => {
    expect(mockRequest).toHaveBeenCalledTimes(1);
    expect(mockRequest).toHaveBeenCalledWith(`/ip/search?ip=${value}`);
  });
  return { getByLabelText, ...rest };
};

it("matches snapshot", () => {
  const { asFragment } = render(<Home />);
  expect(asFragment()).toMatchSnapshot();
});

it("can search for an ip", async () => {
  await renderAndSearch("1.1.1.1");
});

it("can search for the same ip", async () => {
  mockRequest.mockResolvedValue(baseIP);
  const { getByLabelText, queryAllByText } = await renderAndSearch("1.1.1.1");
  fireEvent.change(getByLabelText("IP Address input"), {
    target: { value: baseIP.traits.ipAddress! },
  });
  await wait(() => {
    expect(getByLabelText("IP Address input")).toHaveValue(
      baseIP.traits.ipAddress
    );
  });
  fireEvent.click(getByLabelText("Search IP Address"));
  await wait(() => {
    expect(mockRequest).toHaveBeenCalledTimes(2);
    expect(mockRequest.mock.calls[1][0]).toStrictEqual(
      `/ip/search?ip=${baseIP.traits.ipAddress!}`
    );
  });
  await wait(() => {
    expect(queryAllByText(baseIP.traits.ipAddress!)).toHaveLength(1);
  });
});

it("adds search result to ip menu", async () => {
  mockRequest.mockResolvedValue(baseIP);
  const { getByText } = await renderAndSearch("1.1.1.1");
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
});

it("collapses menu on set current ip on mobile", async () => {
  mediaQueryMock.mockReturnValue(true);
  mockRequest.mockResolvedValue(baseIP);
  const { getByText, queryByText } = await renderAndSearch("1.1.1.1");
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  fireEvent.click(getByText(baseIP.traits.ipAddress!));
  await wait(() => {
    expect(queryByText(baseIP.traits.ipAddress!)).not.toBeInTheDocument();
  });
});

it("does not collapse menu on set current ip on desktop", async () => {
  mediaQueryMock.mockReturnValue(false);
  mockRequest.mockResolvedValue(baseIP);
  const { getByText } = await renderAndSearch("1.1.1.1");
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  fireEvent.click(getByText(baseIP.traits.ipAddress!));
  await wait(() => {
    expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  });
});

it("throws error if latitude not present", async () => {
  mockRequest.mockResolvedValue({
    ...baseIP,
    location: {
      ...baseIP.location,
      latitude: undefined!,
    },
  });
  const { getByText } = await renderAndSearch("1.1.1.1");
  expect(getByText("Unable to locate IP")).toBeInTheDocument();
});

it("removes ip on trash icon click", async () => {
  mockRequest.mockResolvedValue(baseIP);
  const { getByText, queryByText, getByLabelText } = await renderAndSearch(
    "1.1.1.1"
  );
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  fireEvent.click(getByLabelText("remove ip address"));
  await wait(() => {
    expect(queryByText(baseIP.traits.ipAddress!)).not.toBeInTheDocument();
  });
});

it("toggles visibility on eye icon click", async () => {
  getItemMock.mockReturnValue(JSON.stringify([otherIP]));
  mockRequest.mockResolvedValue(baseIP);
  const {
    getByText,
    queryByLabelText,
    getByLabelText,
    getAllByLabelText,
  } = await renderAndSearch("1.1.1.1");
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  fireEvent.click(getAllByLabelText("toggle ip visibility off")[0]);
  await wait(() => {
    expect(queryByLabelText("toggle ip visibility on")).toBeInTheDocument();
  });
  fireEvent.click(getByLabelText("toggle ip visibility on"));
  await wait(() => {
    expect(getAllByLabelText("toggle ip visibility off")).toHaveLength(2);
  });
});

it("populates from localstorage", async () => {
  mockRequest.mockResolvedValue(baseIP);
  getItemMock.mockReturnValue(JSON.stringify([otherIP]));
  const { getByText } = await renderAndSearch("1.2.2.2");
  expect(getByText(otherIP.traits.ipAddress)).toBeInTheDocument();
  expect(getByText(baseIP.traits.ipAddress!)).toBeInTheDocument();
  expect(getItemMock).toHaveBeenCalledTimes(1);
  expect(getItemMock.mock.calls[0][0]).toStrictEqual("allIps");
  expect(setItemMock).toHaveBeenCalledTimes(3);
  expect(setItemMock.mock.calls[2][0]).toStrictEqual("allIps");
  expect(setItemMock.mock.calls[2][1]).toStrictEqual(
    JSON.stringify([baseIP, otherIP])
  );
});
