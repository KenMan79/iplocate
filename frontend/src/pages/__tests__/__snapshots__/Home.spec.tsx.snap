// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  .c11 {
  background: #fff;
  border: 2px solid var(--text-lightest);
  border-radius: 5px;
  padding: 1rem;
  font-size: 1.2rem;
  width: 100%;
  color: var(--text-primary);
  font-family: var(--font-family-mono);
}

.c3 {
  padding: 2rem 1rem;
  margin: 0;
  border-radius: 5px;
  background-color: var(--background-card);
}

.c5 {
  margin: 0;
  padding: 0;
  color: var(--text-primary);
  font-size: 1.8rem;
}

.c6 {
  border-radius: 50%;
  display: block;
  margin-left: 0.5rem;
  font-size: 1.2rem;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-y: auto;
  padding-top: "1rem";
}

.c4 {
  background: none;
  cursor: pointer;
  outline: none;
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
  background: transparent;
}

.c1 .c2 {
  z-index: 1000;
  position: absolute;
  min-width: 400px;
  max-height: 70vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  top: 50%;
  left: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%);
  -ms-transform: translateX(-50%) translateY(-50%);
  transform: translateX(-50%) translateY(-50%);
  -webkit-transition: -webkit-transform 1s ease,top 1s ease,left 1s ease;
  -webkit-transition: transform 1s ease,top 1s ease,left 1s ease;
  transition: transform 1s ease,top 1s ease,left 1s ease;
  box-shadow: 0px 3px 3px -2px rgba(0,0,0,0.2), 0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12);
}

.c1 .c7 {
  -webkit-scrollbar-width: 0;
  -moz-scrollbar-width: 0;
  -ms-scrollbar-width: 0;
  scrollbar-width: 0;
  -ms-overflow-style: none;
}

.c1 .c7::-webkit-scrollbar {
  display: none;
}

.c9 {
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
}

.c9 .c10 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  border-right: none;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-color: var(--text-lighter);
}

.c9 .c10:focus {
  outline: none;
  border-color: var(--text-secondary);
  border-right: none;
}

.c9 button {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  border: none;
  outline: none;
  border-top-right-radius: 5px;
  border-bottom-right-radius: 5px;
  background: var(--text-secondary);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 1.2rem;
  font-size: 18px;
  color: #fff;
  cursor: pointer;
}

.c9 button svg {
  display: block;
}

.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow-y: hidden;
  position: relative;
  width: 100%;
  height: 100%;
}

@media screen and (max-width:450px) {
  .c3 {
    padding: 1rem;
  }
}

@media screen and (max-width:450px) {
  .c1 .c2 {
    min-width: unset;
    top: 10px;
    left: 10px;
    right: 10px;
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      title="IP Locate"
    >
      <div
        class="c2 c3"
      >
        <button
          aria-label="toggle menu expanded"
          class="c4"
        >
          <h2
            class="c5"
          >
            IP Locate
          </h2>
          <span
            class="c6"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
              />
            </svg>
          </span>
        </button>
        <div
          class="c7 c8"
        >
          <form
            class="c9"
          >
            <input
              aria-label="IP Address input"
              class="c10 c11"
              placeholder="IP Address"
              value=""
            />
            <button
              aria-label="Search IP Address"
              type="submit"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                />
              </svg>
            </button>
          </form>
        </div>
      </div>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
